### DISCUSSION TITLE
err.s(3): Error: Constant expression expected

### DISCUSSION THREAD

**Discussion by agsb** [cc65/cc65#1967] [Q&A]
test.s:
;--------------------------

_here_ .set *

;--------------------------

err.s(2): Error: Constant expression expected

Why _here_ is not PC ?

How assign PC to a variable by .set ?

**Comment by mrdudz** [comment]
What do you actually want to achive there?

**Comment by agsb** [comment]
_last_ .set 0
_link_ .set 0

_list_ = *  ; works, _list_ gets PC

.macro link_list
_last_ .set *  ; not work, context is multiplication ?
.word _link_
_link_ .set _last_
; some struct for data
.endmacro
      
I need that _last_ gets the PC at macro call, but seems that * is getting context as multiplication

**Comment by spiro-trikaliotis** [comment]
I think you primary problem can be solved like this:
```
.macro link_list
      .local last
last  .word link
link  .set last
; some struct for data
.endmacro
```

However, this will still not work, because last is not a constant expression, so the line `link .set last` will fail.

**Reply by agsb** [reply]
1
   2 .macro link_list
   3       .local last
   4 last  .word link
   5 link  .set last
   6 ; some struct for data
   7 .endmacro
   8
   9 link_list one
  10


(line numbers only for visual)

err2.s:10: Error: Too many macro parameters
err2.s:10: Error: ':' expected
err2.s:5: Note: Macro was defined here
err2.s:10: Error: Constant expression expected
err2.s:6: Note: Macro was defined here

no work.

**Comment by groessler** [comment]
call me dense, but isn't this the same as

```
.macro link_list
      .local last
last  .word last
; some struct for data
.endmacro
```

**Reply by agsb** [reply]
No.  
the link field must get the previous value of link variable, then link variable get the value of last variable, that is the address of actual link field. repeat while need, and does the link list.

**Reply by groessler** [reply]
> call me dense, but isn't this the same as

I know what linked lists are. I meant, what you wrote (and trying to get compiled) boils down what I wrote.

**Reply by groessler** [reply]
Ah, I see. "link" should be the previous "last", or so...

**Reply by agsb** [reply]
My try is: in  __last_ .set *_ the _*_ is interpreted as multiplication sign not as PC variable.

is a bug.

**Comment by kugelfuhr** [comment]
Reading the documentation helps. First go [here](https://cc65.github.io/doc/ca65.html#ss11.100). Then [follow the link pointing to the full description](https://cc65.github.io/doc/ca65.html#variables) and you will get the error message explained and a suggestion for a possible solution.

The relevant text that explains the error message is

> Since the value of the symbol can change later, it must be possible to evaluate it when used (no delayed evaluation as with normal symbols). So the expression used as the value must be constant.

ca65 generates relocatable output, so the exact PC is unknown at translation time => you get "Constant expression expected".

The first solution is to make the PC constant by adding `.org` somewhere. While obvious and simple, the solution is suboptimal because the code is no longer relocatable.

The second solution is suggested in the docs just below the text explaining the error. Use a variable counter to generate labels. Like this:

```
lcount .set 0          ; Initialize the counter
L0000 = 0

.macro  link_list
.ident(.sprintf("L%04X", lcount + 1)) = *
        .word   .ident (.sprintf ("L%04X", lcount))
        lcount .set lcount + 1
.endmacro
```

**Reply by agsb** [reply]
then I learn that _.set_ is not as _=_

**Comment by agsb** [comment]
Thanks, it works !!!

Em sex., 13 de jan. de 2023 às 11:48, kugelfuhr ***@***.***>
escreveu:

> Reading the documentation helps. First go here
> <https://cc65.github.io/doc/ca65.html#ss11.100>. Then follow the link
> pointing to the full description
> <https://cc65.github.io/doc/ca65.html#variables> and you will get the
> error message explained and a suggestion for a possible solution.
>
> The relevant text that explains the error message is
>
> Since the value of the symbol can change later, it must be possible to
> evaluate it when used (no delayed evaluation as with normal symbols). So
> the expression used as the value must be constant.
>
> ca65 generates relocatable output, so the exact PC is unknown at
> translation time => you get "Constant expression expected".
>
> The first solution is to make the PC constant by adding .org somewhere.
> While obvious and simple, the solution is suboptimal because the code is no
> longer relocatable.
>
> The second solution is suggested in the docs just below the text
> explaining the error. Use a variable counter to generate labels. Like this:
>
> lcount .set 0          ; Initialize the counter
> L0000 = 0
>
> .macro  link_list
> .ident(.sprintf("L%04X", lcount + 1)) = *
>         .word   .ident (.sprintf ("L%04X", lcount))
>         lcount .set lcount + 1
> .endmacro
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cc65/cc65/discussions/1967#discussioncomment-4678577>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR73T2F3T7SFDU7IEV2SBDWSFTL7ANCNFSM6AAAAAATXLZFWU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Slings move continents, a rock at a time.
